### ISSUE TITLE
Boolean equality comparisons should use "IS NOT" instead of "<>"

### ISSUE THREAD

**Issue by dekelpilli** [jkk/honeysql#374]
`x <> true` only returns rows where `x` is false, rather than rows where `x` is false + rows where `x` is null.

Say I have the following table and data:
```sql
create table testing(
    id INT PRIMARY KEY,
    yesno BOOLEAN
);

INSERT INTO testing (id, yesno)
		values(1, TRUE), (2, FALSE), (3, NULL);
```

Say I want to filter based on the yesno column:
```clojure
(-> (hsql/format {:select [:*]
                  :from [:testing]
                  :where [:is-not :yesno true]})
    (first))
=> "SELECT * FROM testing WHERE yesno <> TRUE"
```
The problem is that `SELECT * FROM testing WHERE yesno <> TRUE` would only return the row with ID = 2, whereas I would expect that query to return 2 and 3. In contrast, `select * from testing where yesno is not true` does return both 2 and 3. The query returns the same result (i.e. only id=2) if I used `[:not [:is :yesno true]]`.

I’m using honeysql 2.1.833 and Postgres 13.2.

**Comment by iarenaza** [comment]
According to ANSI-SQL92, any comparison with `NULL` results in an `unknown` result (section 8.2, sub-section "General Rules", paragraph 1.a). That is, neither true, nor false. And according to the ternary logic of ANSI-SQL92 (section 8.12, sub-section "General Rules", paragraph 2), `NOT unknown` results in `unknown` too. So `x <> true` results in `unknown` for the rows where `x` is `NULL`. Again, according to ANSI-SQL92 (section 7.6, sub-section "General Rules", paragraph 1)

> The <search condition> is applied to each row of T. The result of the <where clause> is a table of those rows of T for which the result of the <search condition> is **true**.

(emphasis mine). So it must be explicitly `true` to be taken into account (as opposed to not being `false`, which would also include `unknown`). And `unknown` it not explicitly `true`. So the rows where `x` is `NULL` are discarded.

**Comment by seancorfield** [comment]
@iarenaza That's what I would have expected but @dekelpilli seems to be saying that `WHERE yesno <> TRUE` and `WHERE yesno IS NOT TRUE` do not behave the same, which surprises me.

I was under the impression that `IS` and `=` were synonymous and `IS NOT` and `<>` were synonymous.

**Comment by dekelpilli** [comment]
@seancorfield As @iarenaza said, `x <> true` will give unknown for rows where x is NULL because `NOT unknown` results in `unknown`. This is consistent with my testing and with this comment - https://stackoverflow.com/a/9825992/9854326. On further investigation, I see this behaviour isn't unique to boolean values, but `is/is not` isn't supported for other value types:

```sql
create table testing2(
    id INT PRIMARY KEY,
    num INT
);

INSERT INTO testing2 (id, num)
		values(1, 1), (2, 2), (3, NULL);

select * from testing2 where num = 1; --returns 1
select * from testing2 where num <> 1; --returns 2
select * from testing2 where not(num = 1); --returns 2
select * from testing2 where num is not 1; --syntax error
select * from testing2 where num <> 1 or num is null; --returns 2 and 3
select * from testing2 where num = 1 is not true; --returns 2 and 3
```

**Comment by seancorfield** [comment]
That is, indeed, an interesting wrinkle in how things work. I'll have to think carefully about how to support that without breaking anything else...

**Comment by seancorfield** [comment]
I've confirmed this works on MySQL. Given that `:is` and `:is-not` exist today and are synonyms for `:=` and `:<>` in the grammar, changing them unconditionally is potentially breaking, so this is going to need to be handled via some sort of global flag I think as well as a `format` option.

**Comment by seancorfield** [comment]
Existing code in the wild relies on `[:= :col nil]` and `[:<> :col nil]` so that behavior (special-cased in HoneySQL) needs to be retained -- mapping (in)equality to `IS (NOT) NULL` tests.

I don't know whether existing code relies on the current commutativity regarding `nil` but that might as well be kept too (or perhaps become a lint check?).

I don't know how much code relies on `[:is :col nil]` or `[:is-not :col nil]` but that behavior would not be changed by this -- although `[:is nil :col]` and `[:is-not nil :col]` would either need to be special-cased or just allowed to produce invalid SQL (the latter would be a breaking change).

Just removing the aliasing would:

- ✅  not change `[:is :col nil]` or `[:is-not :col nil]`
- ✅  not change `[:is :col true]` or `[:is :col false]`
- ❌  change the meaning of `[:is-not :col true]` / `[:is-not :col false]` -- those would include `NULL` values which `:<>` does not (this is the core of this particular issue)
- ⚠️  break `[:is :col value]` and `[:is-not :col value]` for non-`nil`, non-Boolean values -- producing invalid SQL (see notes above)
- ⚠️  break `[:is value :col]` and `[:is-not value :col]` for any value (`nil` or non-`nil`) -- producing invalid SQL (could be special-cased and/or flagged by the lint checks)

To properly support Boolean conditions, `[:unknown]` should be added as special syntax so you could test `[:is :col [:unknown]]` or `[:is-not :col [:unknown]]` which is the Boolean version of `nil` essentially. Note: this needs to be explicit because HoneySQL doesn't know about column types -- although using `nil` here to produce `NULL` would work, at least for MySQL.

Just to add further complexity, SQL Server has a setting to change the behavior of equality tests that involve `NULL`: you can configure the database to treat `col = NULL` as `col IS NULL`, rather than producing `UNKNOWN` and therefore not matching `NULL` values. That is not syntax that HoneySQL could currently produce, without resorting to `:raw` to get around the special-casing of `nil` (which code in the wild relies on).

The bottom line is that strict backward compatibility dictates that this must be an opt-in change somehow, although it's questionable how much real world code this would actually break. _I suspect that not many people are actually aware that `:is` and `:is-not` are currently aliased to `:=` and `:<>`?_

**Comment by seancorfield** [comment]
It could also be argued that the current behavior of `[:is-not :col true]` and `[:is-not :col false]` is just broken and changing that behavior would simply be fixing a bug. At least `:is-not` is fairly easy to grep for in source code.

**Comment by seancorfield** [comment]
Data point: using https://grep.app/search?q=%3Ais-not&filter[lang][0]=Clojure does not find `:is-not` used for any other expression than `[:is-not sql-expr nil]` so none of those usages would be broken by this change.

**Comment by seancorfield** [comment]
I have decided to just remove the (incorrect) aliasing of `:is` and `:is-not` and treat this purely as a bug fix. The changelog explicitly notes the change in behavior.
